### ISSUE TITLE
Binary code transforms map creation operators to annihilation operators

### ISSUE THREAD

**Issue by kevinsung** [quantumlib/OpenFermion#251]
Executing the lines
```
ferm_op = FermionOperator('2')
n_modes = 5
qubit_op = binary_code_transform(ferm_op, parity_code(n_modes))
print(qubit_op)
```
yields
```
0.5 [Z1 X2 X3 X4] +
(-0-0.5j) [Y2 X3 X4]
```
However, according to Eq. (19) of arXiv:1208.5986, the second term should have a plus rather than minus sign. @conta877 or @msteudtner do you have any idea why?

**Comment by conta877** [comment]
Yeah I agree with you based on eq 19. 
I'll wait @msteudtner to comment on this.

**Comment by msteudtner** [comment]
Hi, I can take a look at it tomorrow, but I forsee that there isn't much I
can do: the parity code gives isospectral results (we have a test) and I
used the same matrix as in their paper. Are we sure the sign isnt flipped
in eq. 20 and 21 ?

Am 12.03.2018 22:19 schrieb "Isil Ozfidan" <notifications@github.com>:

**Comment by kevinsung** [comment]
I'm quite sure that you're mapping creation operators to annihilation operators and vice versa. This cannot be tested for by looking at the spectrum (particle-hole transformations are just a change of basis). However, if we map all the way to sparse matrices, then we can initialize a state with only mode 1 occupied, and try to annihilate mode 1. This should give the vacuum state, but instead it just gives the zero vector:

```
annihilate_1 = binary_code_transform(FermionOperator('1'), parity_code(2))
annihilate_1_sparse = get_sparse_operator(annihilate_1)

state = jw_configuration_state(occupied_orbitals=[1], n_qubits=2)
result = annihilate_1_sparse * state
print(result.toarray())
```
yields
```
[[0.+0.j]
 [0.+0.j]
 [0.+0.j]
 [0.+0.j]]
```
Note that in this situation the initial state is represented by the string `01` in both the Jordan-Wigner and parity bases.

**Comment by conta877** [comment]
it is isospectral. sign is not flipped in eq21. (the annihilation operator)
This is actually a pretty deep bug - awesome catch @kevinsung 

In the parity operator way it is defined currently the decoder is not the inverse of the encoder because its mod2. 
I believe the way this should be handled is when there is a -1 in a decoder matrix, in linearization one should look at its absolute value to introduce the qubit variable and its sign to add 1. 

I'll push a suggestion that I think @msteudtner should review.

**Comment by kevinsung** [comment]
It seems that you want the inverse over the binary field, and not the inverse over the real numbers. Is there an intrinsic problem with assuming all decoders are input as zero-one matrices (over the binary field)?

**Comment by conta877** [comment]
yeah I'm very wrong - they are still inverse of one another as binary. My fix might not be the right sign fix here then.

**Comment by kevinsung** [comment]
Yes, the `enc_mtx` and `dec_mtx` defined in `parity_code` are indeed inverses modulo 2.

I think the big clue here is that creation operators are getting mapped to annihilation operators and vice versa, so we need to think about what could be causing that. Other than that, the mapping seems perfect.

**Comment by kevinsung** [comment]
It turns out that this issue is not unique to the parity code. Even for the Jordan-Wigner, it is mapping creation operators to annihilation operators and vice versa:
```
ferm_op = FermionOperator('5^')
# Map using canonical Jordan-Wigner
op1 = jordan_wigner(ferm_op)
# Map using code transforms
op2 = binary_code_transform(ferm_op, jordan_wigner_code(6))
print(op1)
print()
print(op2)
```
yields
```
0.5 [Z0 Z1 Z2 Z3 Z4 X5] +
-0.5j [Z0 Z1 Z2 Z3 Z4 Y5]

0.5 [Z0 Z1 Z2 Z3 Z4 X5] +
0.5j [Z0 Z1 Z2 Z3 Z4 Y5]
```
As you can see, the two operators are different. After we fix this, I'd like to see a test like `op1.isclose(op2)` should evaluate to True.

**Comment by msteudtner** [comment]
oh yeah hehe,  I just noticed myself. I have an idea where the problem is. I feel though that I have that one checked already a thousand times. 
Very good catch, @kevinsung !!

**Comment by conta877** [comment]
ok this actually relieves me. its the count variable in transform function.

**Comment by kevinsung** [comment]
This is an interesting example of a bug that can't be caught be checking the spectrums of operators!

**Comment by msteudtner** [comment]
indeed. I can fix it after fixing my jetlag. I'll keep you updated.

**Comment by kevinsung** [comment]
Thank you!

**Comment by conta877** [comment]
#254 fixes the issue. 
I will not work on stuff until after my second cup of :coffee: 

again, this is a great catch @kevinsung

**Comment by kevinsung** [comment]
Closed by #254 .
